File: output/GPT4o-Mini/xCodeEval_remove_4000/C++/Java/1dd8c71fc6620128bda2ac2940bcee8f.java
import java.util.*;
import java.io.*;

public class 1dd8c71fc6620128bda2ac2940bcee8fxcodeeval_processed_4000.json {

    static final int MOD = 1_000_000_007; // 998'244'353
    static final double EPS = 1e-9;
    static final double PI = Math.acos(-1.);
    static long LLINF = 18_E + 10;
    static int INF = 9_E + 10;

    static final int MAXN = 6_E + 6;
    static final int N = 5_E + 5;

    static int[] cnt = new int[MAXN];

    static class Fenwick {
        private int[] fw;

        Fenwick(int n) {
            fw = new int[n + 1];
        }

        void update(int i, int d) {
            while (i < fw.length) {
                fw[i] = Math.max(fw[i], d);
                i = (i | (i + 1));
            }
        }

        int get(int n) {
            int result = 0;
            while (n >= 0) {
                result = Math.max(result, fw[n]);
                n = (n & (n + 1)) - 1;
            }
            return result;
        }

        int get(int l, int r) {
            return get(r) - get(l - 1);
        }
    }

    static void solution(Scanner cin, PrintWriter cout) {
        int n = cin.nextInt();
        int m = cin.nextInt();
        for (int i = 1; i <= n; i++) {
            int l = cin.nextInt();
            int r = cin.nextInt();
            cnt[l]++;
            cnt[r + 1]--;
        }

        for (int i = 1; i <= m; i++) {
            cnt[i] += cnt[i - 1];
        }

        Fenwick fw_l = new Fenwick(100001);
        int[] l = new int[MAXN];
        for (int i = 1; i <= m; i++) {
            l[i] = fw_l.get(cnt[i]) + 1;
            fw_l.update(cnt[i], l[i]);
        }

        Fenwick fw_r = new Fenwick(100001);
        int[] r = new int[MAXN];
        for (int i = m; i >= 1; i--) {
            r[i] = fw_r.get(cnt[i]) + 1;
            fw_r.update(cnt[i], r[i]);
        }

        int ans = 0;
        for (int i = 1; i <= m; i++) {
            ans = Math.max(ans, l[i] + r[i] - 1);
        }

        cout.println(ans);
    }

    public static void main(String[] args) {
        Scanner cin = new Scanner(System.in);
        PrintWriter cout = new PrintWriter(System.out);

        int queries = 1;

        for (int test_case = 1; test_case <= queries; test_case++) {
            solution(cin, cout);
        }

        cout.flush();
        cin.close();
    }
}

File: output/GPT4o-Mini/xCodeEval_remove_4000/C++/Python/1dd8c71fc6620128bda2ac2940bcee8f.py
import sys
from collections import defaultdict
from itertools import accumulate

MOD = 1_000_000_007
EPS = 1e-9
PI = 3.141592653589793
LLINF = 18 * 10**10
INF = 9 * 10**10

MAXN = 6 * 10**6
N = 5 * 10**5

cnt = [0] * (MAXN + 1)

class Fenwick:
    def __init__(self, n):
        self.fw = [0] * (n + 1)

    def update(self, i, d):
        while i < len(self.fw):
            self.fw[i] = max(self.fw[i], d)
            i |= (i + 1)

    def get(self, n):
        result = 0
        while n >= 0:
            result = max(result, self.fw[n])
            n = (n & (n + 1)) - 1
        return result

    def range_get(self, l, r):
        return self.get(r) - self.get(l - 1)

def solution(cin, cout):
    n, m = map(int, cin.readline().split())
    for _ in range(n):
        l, r = map(int, cin.readline().split())
        cnt[l] += 1
        cnt[r + 1] -= 1

    cnt[1:] = accumulate(cnt[1:])

    fw_l = Fenwick(100001)
    l = [0] * (m + 1)
    for i in range(1, m + 1):
        l[i] = fw_l.get(cnt[i]) + 1
        fw_l.update(cnt[i], l[i])

    fw_r = Fenwick(100001)
    r = [0] * (m + 1)
    for i in range(m, 0, -1):
        r[i] = fw_r.get(cnt[i]) + 1
        fw_r.update(cnt[i], r[i])

    ans = 0
    for i in range(1, m + 1):
        ans = max(ans, l[i] + r[i] - 1)

    cout.write(f"{ans}\n")

def main():
    input = sys.stdin.read
    data = input().splitlines()
    cin = iter(data)
    cout = sys.stdout

    queries = 1
    for _ in range(queries):
        solution(cin, cout)

if __name__ == "__main__":
    main()
